kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java

```diff
@@ -47,15 +60,28 @@ public void channelReadComplete(ChannelHandlerContext ctx) {
 
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        cause.printStackTrace();
         ctx.close();
     }
 
-    /**
-     * A future that is completed when the response is received
-     * @return on response future
-     */
-    public CompletableFuture<DecodedResponseFrame<?>> getOnResponseFuture() {
-        return onResponse;
+    public CompletableFuture<DecodedResponseFrame<?>> sendRequest(DecodedRequestFrame<?> decodedRequestFrame) {
+        queue.addLast(decodedRequestFrame);
+        processPendingWrites();
+        return decodedRequestFrame.getResponseFuture();
+    }
+
+    private void processPendingWrites() {
+        ctx.executor().execute(() -> {
+            if (ctx.channel().isActive()) {
+                while (queue.peek() != null) {
+                    DecodedRequestFrame<?> msg = queue.removeFirst();
+                    ctx.writeAndFlush(msg).addListener(c -> {
+                        if (msg.body() instanceof ProduceRequest && ((ProduceRequest) msg.body()).acks() == 0) {
+                            msg.getResponseFuture().complete(null);
+                        }
```

Comment:
Can we extract this special case somehow? Maybe it should be:
```suggestion
                        if (!msg.ackRequired) {
                            msg.getResponseFuture().complete(null);
                        }
```

Doesn't need to happen as part of this PR.

Reply:
Done.